flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -162,34 +162,98 @@ public Dispatcher(
+            DispatcherId fencingToken,
+            Collection<JobGraph> recoveredJobs,
+            Collection<JobResult> recoveredDirtyJobs,
+            Configuration configuration,
+            HighAvailabilityServices highAvailabilityServices,
+            GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever,
+            HeartbeatServices heartbeatServices,
+            BlobServer blobServer,
+            FatalErrorHandler fatalErrorHandler,
+            JobGraphWriter jobGraphWriter,
+            JobResultStore jobResultStore,
+            JobManagerMetricGroup jobManagerMetricGroup,
+            @Nullable String metricServiceQueryAddress,
+            Executor ioExecutor,
+            HistoryServerArchivist historyServerArchivist,
+            ExecutionGraphInfoStore executionGraphInfoStore,
+            JobManagerRunnerFactory jobManagerRunnerFactory,
+            DispatcherBootstrapFactory dispatcherBootstrapFactory,
+            DispatcherOperationCaches dispatcherOperationCaches,
+            JobManagerRunnerRegistry jobManagerRunnerRegistry)
+            throws Exception {
         super(rpcService, RpcServiceUtils.createRandomName(DISPATCHER_NAME), fencingToken);
-        checkNotNull(dispatcherServices);
         assertRecoveredJobsAndDirtyJobResults(recoveredJobs, recoveredDirtyJobs);
 
-        this.configuration = dispatcherServices.getConfiguration();
-        this.highAvailabilityServices = dispatcherServices.getHighAvailabilityServices();
-        this.resourceManagerGatewayRetriever =
-                dispatcherServices.getResourceManagerGatewayRetriever();
-        this.heartbeatServices = dispatcherServices.getHeartbeatServices();
-        this.blobServer = dispatcherServices.getBlobServer();
-        this.fatalErrorHandler = dispatcherServices.getFatalErrorHandler();
-        this.jobGraphWriter = dispatcherServices.getJobGraphWriter();
-        this.jobResultStore = dispatcherServices.getJobResultStore();
-        this.jobManagerMetricGroup = dispatcherServices.getJobManagerMetricGroup();
-        this.metricServiceQueryAddress = dispatcherServices.getMetricQueryServiceAddress();
-        this.ioExecutor = dispatcherServices.getIoExecutor();
+        this.configuration = checkNotNull(configuration);
+        this.highAvailabilityServices = checkNotNull(highAvailabilityServices);
+        this.resourceManagerGatewayRetriever = checkNotNull(resourceManagerGatewayRetriever);
+        this.heartbeatServices = checkNotNull(heartbeatServices);
+        this.blobServer = checkNotNull(blobServer);
+        this.fatalErrorHandler = checkNotNull(fatalErrorHandler);
+        this.jobGraphWriter = checkNotNull(jobGraphWriter);
+        this.jobResultStore = checkNotNull(jobResultStore);
+        this.jobManagerMetricGroup = checkNotNull(jobManagerMetricGroup);
+        this.metricServiceQueryAddress = metricServiceQueryAddress;
+        this.ioExecutor = checkNotNull(ioExecutor);
```

Comment:
I'm not really a big fan of these safeguards.

- It's hard to keep them in sync when new code is added
- We should be able to enforce this in compile time simply by checking annotations (eg. https://github.com/uber/NullAway)
- It adds noise

What was the reason for adding these checks?